apis/v1alpha2/gateway_types.go

```diff
@@ -0,0 +1,826 @@
+	// determined in order of the following criteria:
+	//
+	// * The most specific match as defined by the Route type.
+	// * The oldest Route based on creation timestamp. For example, a Route with
+	//   a creation timestamp of "2020-09-08 01:02:03" is given precedence over
+	//   a Route with a creation timestamp of "2020-09-08 01:02:04".
+	// * If everything else is equivalent, the Route appearing first in
+	//   alphabetical order (namespace/name) should be given precedence. For
+	//   example, foo/bar is given precedence over foo/baz.
+	//
+	// All valid rules within a Route attached to this Listener should be
+	// implemented. Invalid Route rules can be ignored (sometimes that will mean
+	// the full Route). If a Route rule transitions from valid to invalid,
+	// support for that Route rule should be dropped to ensure consistency. For
+	// example, even if a filter specified by a Route rule is invalid, the rest
+	// of the rules within that Route should still be supported.
+	//
+	// Support: Core
+	// +kubebuilder:default={namespaces:{from: Same}}
+	// +optional
+	AllowedRoutes *AllowedRoutes `json:"allowedRoutes,omitempty"`
+}
+
+// ProtocolType defines the application protocol accepted by a Listener.
+// Implementations are not required to accept all the defined protocols.
+// If an implementation does not support a specified protocol, it
+// should raise a "Detached" condition for the affected Listener with
+// a reason of "UnsupportedProtocol".
+//
+// Core ProtocolType values are listed in the table below.
+//
+// Implementations can define their own protocols if a core ProtocolType does not
+// exist. Such definitions must use prefixed name, such as
+// `mycompany.com/my-custom-protocol`. Un-prefixed names are reserved for core
+// protocols. Any protocol defined by implementations will fall under custom
+// conformance.
+type ProtocolType string
+
+const (
+	// Accepts cleartext HTTP/1.1 sessions over TCP. Implementations MAY also
+	// support HTTP/2 over cleartext. If implementations support HTTP/2 over
+	// cleartext on "HTTP" listeners, that MUST be clearly documented by the
+	// implementation.
+	HTTPProtocolType ProtocolType = "HTTP"
+
+	// Accepts HTTP/1.1 or HTTP/2 sessions over TLS.
+	HTTPSProtocolType ProtocolType = "HTTPS"
+
+	// Accepts TLS sessions over TCP.
+	TLSProtocolType ProtocolType = "TLS"
+
+	// Accepts TCP sessions.
+	TCPProtocolType ProtocolType = "TCP"
+
+	// Accepts UDP packets.
+	UDPProtocolType ProtocolType = "UDP"
+)
+
+// GatewayTLSConfig describes a TLS configuration.
+type GatewayTLSConfig struct {
+	// Mode defines the TLS behavior for the TLS session initiated by the client.
+	// There are two possible modes:
+	//
+	// - Terminate: The TLS session between the downstream client
+	//   and the Gateway is terminated at the Gateway. This mode requires
+	//   certificateRefs to be set and contain at least one element.
+	// - Passthrough: The TLS session is NOT terminated by the Gateway. This
```

Comment:
I 100% agree this is an important use case but I don't think its a separate tls option in the API. The upstream configuration and downstream configuration is separate. For a given incoming request, we can either terminate the TLS or pass it through. Then for a given outgoing request, we can either send it as is or encap with TLS.

What we do on the outgoing request side should not be coupled here, just like the outgoing protocol is not (for example, terminate HTTP/2 and send HTTP/1.1 upstream).

Note this also means you can passthrough AND encap in TLS. A bit obscure for other cases, but this can be used for service mesh mTLS when the application itself still expects a TLS connection.

Reply:
Agree with everything @howardjohn said above. @khenidak do you think that makes sense? Would clarifying this distinction in the comments here be sufficient?